File: API/_archive/Protobuf/python/usecase4_deep_copy.py
# import models_pb2

class MockUser:
    def __init__(self, id, name):
        self.id = id
        self.name = name
    def SerializeToString(self):
        return b"mock"
    def ParseFromString(self, data):
        self.id = 1
        self.name = "Alice"

def clone_user(original) -> MockUser:
    # cloned = models_pb2.User()
    cloned = MockUser(0, "")
    cloned.ParseFromString(original.SerializeToString())
    return cloned

if __name__ == "__main__":
    user = MockUser(1, "Alice")
    cloned = clone_user(user)
    print(f"Original ID: {id(user)}, Cloned ID: {id(cloned)}")

File: API/_archive/Protobuf/golang/usecase4_deep_copy.go
package main

import (
	"fmt"
	// "google.golang.org/protobuf/proto"
)

type User struct { Id int32; Name string; Email string }
func (u *User) ProtoReflect() {} // conceptual

func cloneUser(original *User) *User {
	// clone := proto.Clone(original).(*pb.User)
	clone := &User{Id: original.Id, Name: original.Name, Email: original.Email}
	return clone
}

func main() {
	user := &User{Id: 1, Name: "Alice"}
	cloned := cloneUser(user)
	fmt.Printf("Original: %p, Cloned: %p\n", user, cloned)
}
